### PR TITLE
Refactored StickyPartitionAssignmentStrategy and implemented task estimation logic in LoadBasedPartitionAssignmentStrategy

### PR DESCRIPTION
**Pull request by jzakaryan**
- Removed `DatastreamClusterResolver` and refactored `ThroughputInfoProvider`. Now it's up to `ThroughputInfoProvider` to find source Kafka cluster and source fabric.
- Refactored `StickyPartitionAssignmentStrategy` to break down the task count validation method into a method which just retrieves the task count estimate and a second one which does the validation and throws the exception.
- Added `_isPartitionNumBasedTaskCountEstimationEnabled` and `_isThroughputBasedTaskCountEstimationEnabled` to be able to toggle them individually. These will be converted to proper configs in one the next PRs.
- Implemented logic for task count estimation in `LoadBasedPartitionAssignmentStrategy`.
- Implemented unit tests for `LoadBasedPartitionAssignmentStrategy`

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/providers/FileBasedPartitionThroughputProvider.java [hunk 1]
```diff
@@ -49,6 +51,11 @@ public ClusterThroughputInfo getThroughputInfo(String clusterName) {
     return readThroughputInfoFromFile(partitionThroughputFile, clusterName);
   }
 
+  @Override
+  public ClusterThroughputInfo getThroughputInfo(DatastreamGroup datastreamGroup) {
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/providers/FileBasedPartitionThroughputProvider.java:55]: Setting this to NotImplemented means, any test added in LoadBasedPartitionAssignmentStrategy is not using the throughput information. Since, this will throw exception and the strategy after retry will switch too default mode.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+import static org.mockito.Mockito.times;
+
+
+/**
+ * Tests for {@link LoadBasedPartitionAssignmentStrategy}
+ */
+@Test
+public class TestLoadBasedPartitionAssignmentStrategy {
+  private EmbeddedZookeeper _embeddedZookeeper;
+  private ZkClient _zkClient;
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:56]: Add tests for
a) Regular case with assignPartitions with FileBasedThroughputProvider.
b) Add tests for partition assignment when new topics get created, meaning the existing assignment is not empty and there are new topics.
c) Look at tests for StickyMulticastStrategy and add them here.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 3]
```diff
@@ -44,28 +45,28 @@
...
   private final int _throughputInfoFetchRetryPeriodMs;
 
+  // TODO Make this a config
+  private final boolean _enableThroughputBasedTaskCountEstimation = true;
+
 
   /**
    * Creates an instance of {@link LoadBasedPartitionAssignmentStrategy}
    */
   public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughputProvider,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:60]: One of the big missing piece for me is the metrics and alerts in this class. Please consider how you intend to debug the live setup when some issue happens. Some of the interesting pieces can be
a) How many times ThroughputProvider is invoked. This metric can go to Pinot specific Throughput Provider.
b) Current code will not generate alert for SREs to increase the maxTasks, as there is more demand based on throughput. `updateOrRegisterElasticTaskAssignmentMetrics` is currently tightly coupled with partition count, which will not work for throughput.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 2 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:56]: Tests for assignment itself are in a separate file. There I already cover cases when existing assignment is not empty. I don't think there's a reason to duplicate those tests in the strategy. My goal with strategy tests was mostly to cover fallback mechanisms and zk updates.

File: datastream-server/src/main/java/com/linkedin/datastream/server/providers/FileBasedPartitionThroughputProvider.java (hunk 1 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/providers/FileBasedPartitionThroughputProvider.java:55]: The tests for assignment logic are in a separate file. Strategy tests don't use `FileBasedPartitionThroughputProvider` and instead use a mock when necessary.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 4]
```diff
@@ -77,59 +78,84 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
-    // Calculating unassigned partitions
-    List<String> unassignedPartitions = new ArrayList<>();
+    // Estimating number of tasks needed
+    int numTasksNeeded = taskCount;
+    if (isPartitionNumBasedTaskCountEstimationEnabled(datastreamGroup)) {
+      numTasksNeeded = getTaskCountEstimateBasedOnNumPartitions(datastreamPartitions, taskCount);
+      updateOrRegisterElasticTaskAssignmentMetrics(datastreamPartitions, taskCount);
+    }
+
+    if (isThroughputBasedTaskCountEstimationEnabled()) {
```
[the remark this answers is not in the document]
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:125]: I'm not convinced that we should couple assignment with estimation. Let's discuss this offline.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 5]
```diff
@@ -77,7 +77,9 @@
   private static final Integer DEFAULT_PARTITION_FULLNESS_FACTOR_PCT = 75;
   private static final DynamicMetricsManager DYNAMIC_METRICS_MANAGER = DynamicMetricsManager.getInstance();
 
-  private final boolean _enableElasticTaskAssignment;
+  protected final boolean _enableElasticTaskAssignment;
+  protected final boolean _enablePartitionNumBasedTaskCountEstimation = true;
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:81]: Agreed to move this boolean to `LoadBasedPartitionAssignmentStrategy`.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 2 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:56]: The following tests cases were added:

- throughput provider is not called when doing incremental assignment.
- maxTasks config is honored

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 6]
```diff
@@ -77,56 +78,79 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
-    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
-    try {
-      partitionThroughputInfo = fetchPartitionThroughputInfo();
-    } catch (RetriesExhaustedException ex) {
-      LOG.warn("Attempts to fetch partition throughput timed out. Falling back to regular partition assignment strategy");
+    // For throughput based partition-assignment to kick in, the following conditions must be met:
+    //   (1) Elastic task assignment must be enabled through configuration
+    //   (2) Throughput-based task assignment must be enabled through configuration
+    if (!isElasticTaskAssignmentEnabled(datastreamGroup) || !_enableThroughputBasedPartitionAssignment) {
+      LOG.info("Throughput based elastic task assignment not enabled. Falling back to sticky partition assignment.");
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:86]: nit: display both the condition values for debugging.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 7]
```diff
@@ -77,56 +78,79 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
+    // Calculating unassigned partitions
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
 
-    // TODO Get task count estimate and perform elastic task count validation
-    // TODO Get task count estimate based on throughput and pick a winner
-    LoadBasedTaskCountEstimator estimator = new LoadBasedTaskCountEstimator(_taskCapacityMBps, _taskCapacityUtilizationPct);
-    int maxTaskCount = estimator.getTaskCount(clusterThroughputInfo, Collections.emptyList(), Collections.emptyList());
-    LOG.info("Max task count obtained from estimator: {}", maxTaskCount);
+    ClusterThroughputInfo clusterThroughputInfo = new ClusterThroughputInfo(StringUtils.EMPTY, new HashMap<>());
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:104]: replace `new HashMap<>()` with `Collections.empty_map()`

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 8]
```diff
@@ -77,56 +78,79 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
+    AtomicInteger attemptNum = new AtomicInteger(0);
     return PollUtils.poll(() -> {
       try {
-        return _throughputProvider.getThroughputInfo();
+        return _throughputProvider.getThroughputInfo(datastreamGroup);
       } catch (Exception ex) {
-        // TODO print exception and retry count
-        LOG.warn("Failed to fetch partition throughput info.");
+        attemptNum.set(attemptNum.get() + 1);
+        LOG.warn("Failed to fetch partition throughput info on attempt {}.", attemptNum.get(), ex);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:153]: This LOG.warn call is going to `public void warn(String format, Object arg1, Object arg2);` instead of `public void warn(String msg, Throwable t);`. This warn may not print.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 9]
```diff
@@ -77,7 +77,8 @@
   private static final Integer DEFAULT_PARTITION_FULLNESS_FACTOR_PCT = 75;
   private static final DynamicMetricsManager DYNAMIC_METRICS_MANAGER = DynamicMetricsManager.getInstance();
 
-  private final boolean _enableElasticTaskAssignment;
+  protected final boolean _enableElasticTaskAssignment;
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:80]: revert this back to `private` from `protected`

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 10]
```diff
@@ -583,25 +588,32 @@ protected void performElasticTaskCountValidation(DatastreamGroupPartitionsMetada
     int totalPartitions = datastreamPartitions.getPartitions().size();
     int numTasksNeeded = (totalPartitions / allowedPartitionsPerTask)
         + (((totalPartitions % allowedPartitionsPerTask) == 0) ? 0 : 1);
+    LOG.info("Number of tasks needed: {}, total task count: {}", numTasksNeeded, totalTaskCount);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:591]: This log will be very useful to debug production issues. Add Datastream information here for debugging. This may not be a big deal while debugging using logs, but with Kusto, this log will not make complete sense. If it is too much to get datastream name info in this, you can summarize all the estimators and final task count in the caller, so that we can use Kusto to look for this log.

Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:591]: Not you fault, but looks like there are bunch of places in elastic task count path where datastream group info will be useful in the log.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 11]
```diff
@@ -583,25 +588,32 @@ protected void performElasticTaskCountValidation(DatastreamGroupPartitionsMetada
...
     }
-    LOG.info("Number of tasks needed: {}, total task count: {}", numTasksNeeded, totalTaskCount);
+    return numTasks;
+  }
+
+  protected void updateNumTasksAndForceTaskCreation(DatastreamGroupPartitionsMetadata datastreamPartitions,
+      int numTasksNeeded, int actualNumTasks) {
+    createOrUpdateNumTasksForDatastreamInZK(datastreamPartitions.getDatastreamGroup().getTaskPrefix(), numTasksNeeded);
+    setTaskCountForDatastreamGroup(datastreamPartitions.getDatastreamGroup().getTaskPrefix(), numTasksNeeded);
+    int totalPartitions = datastreamPartitions.getPartitions().size();
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:610]: nit: Feel free to ignore: this local variable is not really needed.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+
+
+/**
+ * Tests for {@link LoadBasedPartitionAssignmentStrategy}
+ */
+@Test
+public class TestLoadBasedPartitionAssignmentStrategy {
+  private EmbeddedZookeeper _embeddedZookeeper;
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:55]: convert this to local variable.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    Optional<ZkClient> zkClient = Optional.empty();
+
+    LoadBasedPartitionAssignmentStrategy strategy = new LoadBasedPartitionAssignmentStrategy(mockProvider,
+        maxTasks, imbalanceThreshold, maxPartitionPerTask, enableElasticTaskAssignment, partitionsPerTask,
+        partitionFullnessFactorPct, taskCapacityMBps, taskCapacityUtilizationPct, throughputInfoFetchTimeoutMs,
+        throughputInfoFetchRetryPeriodMs, zkClient, _clusterName);
+
+    Datastream ds1 = DatastreamTestUtils.createDatastreams(DummyConnector.CONNECTOR_TYPE, "ds1")[0];
+    ds1.getSource().setPartitions(0);
+    ds1.getMetadata().put(DatastreamMetadataConstants.TASK_PREFIX, DatastreamTaskImpl.getTaskPrefix(ds1));
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:91]: I don't think elasticTaskCount is enabled for the datastream since minTask is not configured on the datastream. To verify, add check if `isElasticTaskAssignmentEnabled` return true for the datastream group.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  @BeforeMethod
+  public void setup() throws IOException {
+    DynamicMetricsManager.createInstance(new MetricRegistry(), "TestStickyPartitionAssignment");
+    _clusterName = "testCluster";
+    _embeddedZookeeper = new EmbeddedZookeeper();
+    String zkConnectionString = _embeddedZookeeper.getConnection();
+    _embeddedZookeeper.startup();
+    _zkClient = new ZkClient(zkConnectionString);
+  }
+
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:68]: One test left is to mock `fetchPartitionThroughputInfo` to throw Exception and add the expected behavior in the tests.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+  @Test
+  public void fallbackToBaseClassWhenElasticTaskAssignmentDisabledTest() {
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:70]: nit: Feel free to ignore this: I just realized that you use Test at the end of function name, while rest of the code-base test names start with Test.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 14 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:68]: What do I assert in that case? How do I test that the `super.assignPartitions(...)` was called without exposing the innards of `StickyPartitionAssignmentStrategy` with `@VisibleForTesting`?

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 13 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:91]: This is the test case for when elastic task assignment is disabled. I don't need it to be enabled.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 10 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java]: Added datastream group name to the highlighted log. We can tackle other log improvements in a separate pull request.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java (hunk 6 above)
Comment (vmaheshw) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:87]: merge the logs.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 16]
```diff
@@ -78,6 +78,7 @@
   private static final DynamicMetricsManager DYNAMIC_METRICS_MANAGER = DynamicMetricsManager.getInstance();
 
   private final boolean _enableElasticTaskAssignment;
+
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:81]: nit: empty line

Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:81]: nit: empty line

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 14 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:67]: One thing I can think of is you move the assignor code to another method exposed with VisibleForTesting and ensure that method is not called, but assignPartitions return new assignment

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 13 above)
Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:90]: Sorry. You should add check if `isElasticTaskAssignmentEnabled` returns false as well.

I'm not able to find the test where elasticTaskAssignment is enabled and throughputProvider call counts are verified.

Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:87]: - `isElasticTaskAssignmentEnabled()` is tested in `fallbackToBaseClassWhenElasticTaskAssignmentDisabledTest(..)` line 95.

- The assertion that `throughputProvider` is called to fetch the data is tested in `updatesNumTasksAndThrowsExceptionWhenNoSufficientTasksTest` line 183.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 14 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:64]: @vmaheshw I don't want this test case to impact the code that much. In order to test whether assignment code is called or not I either need to convert the Assigner to an interface (to mock it and then verify call counts), or save some state in the instance (i.e. something like `assignmentPerformed`) and test against that.

Comment (vmaheshw) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:64]: Why? If you spy the strategy object itself and verify the method call that you extract, it will work.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java [hunk 17]
```diff
@@ -138,14 +140,23 @@ public LoadBasedPartitionAssignmentStrategy(PartitionThroughputProvider throughp
...
-        currentAssignment, unassignedPartitions, datastreamPartitions);
-    LOG.info("new assignment info, assignment: {}", newAssignment);
+    Map<String, Set<DatastreamTask>> newAssignment = doAssignment(clusterThroughputInfo, currentAssignment,
+        unassignedPartitions, datastreamPartitions);
     partitionSanityChecks(newAssignment, datastreamPartitions);
     return newAssignment;
   }
 
+  @VisibleForTesting
+  Map<String, Set<DatastreamTask>> doAssignment(ClusterThroughputInfo clusterThroughputInfo, Map<String,
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:150]: nit: `Map<String,` should be on new line

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java [hunk 18]
```diff
@@ -100,6 +106,45 @@ public void fallbackToBaseClassWhenElasticTaskAssignmentDisabledTest() {
...
+    Optional<Integer> taskCapacityMBps = Optional.of(5);
+    Optional<Integer> taskCapacityUtilizationPct = Optional.of(90);
+    Optional<Integer> throughputInfoFetchTimeoutMs = Optional.of(1000);
+    Optional<Integer> throughputInfoFetchRetryPeriodMs = Optional.of(200);
+    Optional<ZkClient> zkClient = Optional.of(_zkClient);
+
+    LoadBasedPartitionAssignmentStrategy strategy = Mockito.spy(new LoadBasedPartitionAssignmentStrategy(mockProvider,
+        maxTasks, imbalanceThreshold, maxPartitionPerTask, enableElasticTaskAssignment, partitionsPerTask,
+        partitionFullnessFactorPct, taskCapacityMBps, taskCapacityUtilizationPct, throughputInfoFetchTimeoutMs,
+        throughputInfoFetchRetryPeriodMs, zkClient, _clusterName));
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:128]: nit: You can extract the strategy creation option to a separate method to avoid duplicate code. You can use Intellij utility to do the refactoring for you.

**Review by vmaheshw — APPROVED** [review]
2 nits.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java (hunk 18 above)
Comment (jzakaryan) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestLoadBasedPartitionAssignmentStrategy.java:128]: I'd prefer to leave it as it is. If I move the strategy creation in a separate method, I'll have to pass a lot of arguments in each call and it's prone to errors.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java (hunk 3 above)
Comment (jzakaryan) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java:62]: Metrics/alerts will be published in a separate PR.

**Review by vmaheshw — APPROVED** [review]

**Review by shrinandthakkar — APPROVED** [review]
